Block Toolbar: Move the block toolbar to the editor's header

youknowriad opened the pull request:
## Description
refs #2983

This PR moves the block's toolbar to the editor's header.

<img width="1001" alt="screen shot 2017-10-12 at 09 16 10" src="https://user-images.githubusercontent.com/272444/31486068-0a7cebbc-af2e-11e7-84d9-3a9ddef4dc82.png">

## Checklist

- [x] Move the toolbar.
- [x] Always show the toolbar when the block is selected.
- [x] Unify header buttons style
- [ ] Find a better place for the permalink 
- [ ] Address the classic block's toolbar (needs design probably)
- [x] Mobile
- [x] Accessibility (part of this will be addressed by #2960

youknowriad committed:
Block Toolbar: Move the block toolbar to the editor's header

afercia commented:
Copying and pasting the accessibility items from #2960 

1
Screen readers, especially the Windows ones (e.g. NVDA, JAWS) work in two different modes: "browse mode" and "forms mode". In "browse mode" they intercept key presses for their own goals and don't pass them to the browser. As a result, browsers are unaware of JS events on non-interactive elements (in our case the `BlockToolbar`). Only when screen readers switch to "forms mode", e.g. when the users lands on a form field, they pass key presses to the browser. They do the same on non-interactive elements only if they have some special ARIA roles. Only in this case JS events work on non-interactive elements.
In a few words: arrows navigation doesn't work with NVDA and JAWS (and seems it's buggy with Safari/VoiceOver too) because the browser doesn't know a keyboard event happened on the toolbar. To fix this, the toolbar needs an ARIA `role="toolbar"`. Given the current markup, this would be a bit tricky, see below.

2
Toolbars need to have a "flat" structure, just a `div` with controls inside. Not a list.
I understand this is a bit tricky to address because the toolbars are actually composed by multiple toolbars, each one using a list. I've mentioned this issue a while ago in a conversation with @aduth . This should be addressed, and maybe #2148 would be the proper place.
**Update:** the toolbars won't be lists any longer if #3001 gets merged. However, there should be just one wrapper with `role="toolbar"` that contains all the controls. To my understanding, this would be easier with #3001 because the wrapper would contain just divs (and the controls). Will need testing with Windows screen readers.

3
To meet the ARIA spec and the ARIA Authoring Practices, navigation through the toolbar controls should happen with arrow keys only (no Tab). See https://www.w3.org/TR/wai-aria-practices/#toolbar

4
The toolbar should be labeled with an `aria-label` attribute. Worth noting the `role="toolbar"` already makes screen readers announce "tool bar" so the `aria-label` should say just something like "Formatting".

jasmussen commented:
Thank you for working on this. 🏅

I think this is a great start. I think unifying the button style should be done separately, I have on my todo to look at focus styles again, given all that's changed and what we've learned. I have some ideas here, but I think this is probably fine for now. 

The permalink should probably live in the Status menu of the sidebar, and if/when we merge the publish dropdown, there as well. 

Can we keep the Classic blocks docked toolbar? It works pretty well there, and it is a rather unique block compared to all the others. 

I would love to jump in and fix mobile in this one — I hope to add a breakpoint that makes the toolbar stacked below the editor bar — but perhaps it's best I jump in once some of afercia's notes are addressed?

youknowriad commented:
Rebase this to fix the accessibility issues:

 - The toolbar has arrow navigation
 - The toolbar has in/out shortcuts
 - The toolbar has the toolbar role
 - The toolbar is a `div` and no `ul`s anymore
 - The toolbar has a label (probably not a great one but "Formatting" is not generic enough)

jasmussen commented:
I will take a look at the responsiveness/mobile on monday. Thank you Riad!

jasmussen commented:
I'm looking at some polish here but for now I'm seeing this JS warning when setting focus on a text block:

<img width="618" alt="screen shot 2017-10-16 at 11 21 49" src="https://user-images.githubusercontent.com/1204802/31604685-5ae48ae0-b264-11e7-8ba6-c1105c39696f.png">

youknowriad commented:
@jasmussen yep, this should be fixed in master (just rebase)

Commit:
Make the toolbar stack on mobile

jasmussen commented:
I pushed a tweak so that the toolbar stacks on the mobile breakpoint.

But it's not perfect. Long term we should work to lower the breakpoint that causes these to stack, right now they stack already at 960. 

Short term we should think about doing some polish to how the "mobile controls" work. Right now the "mobile controls" are basically the mover, and the now-in-dropdown Settings, Delete and HTML mode. These are now in 2 levels of drawers that all look weird. 

I see two ways we can fix this:

1. Drop the current implementation we have for handling the mover and trash on mobile, and instead move these to the ellipsis dropdown. Then we can put the ellipsis dropdown in the quick toolbar on mobile.

2. Start thinking about https://github.com/WordPress/gutenberg/issues/2898, and how that could potentially solve both the issue of the mover/trash/settings, as well as any item overflow in the quick toolbar.

jasmussen commented:
There's also an issue with the fact that the toolbar disappears entirely when nothing is selected. This means the stacked toolbar overlaps the scrollbar. Perhaps worth thinking about showing a grayed text block toolbar when nothing is selected, or something in that vein.

karmatosed commented:
A minor issue with Hello Dolly and smaller screen sizes:

![2017-10-17 at 12 56](https://user-images.githubusercontent.com/253067/31661364-b13e9e0e-b33a-11e7-9e36-3c245e23f1b6.png)

youknowriad commented:
Rebased this and fixed some issues with the mobile toolbar. Do you think we could merge this soon in the weekend to discover any drawbacks before the next release?

jasmussen commented:
I think it would be good to get this in soon, yes. Nice fixes. 

I wish the stacked toolbar didn't _cover_ the scrollable area, but instead pushed it down:

![screen shot 2017-10-26 at 09 34 45](https://user-images.githubusercontent.com/1204802/32040432-0d028532-ba31-11e7-8a54-c1928f56b9d0.png)

However I feel like when I pushed fixes to this branch, I tried addressing that and running into issues. Perhaps that if you have nothing selected, the stacked toolbar disappears entirely:

![screen shot 2017-10-26 at 09 38 20](https://user-images.githubusercontent.com/1204802/32040562-93dab71e-ba31-11e7-941b-28cc5ca5dc54.png)

Perhaps we should embrace the toolbar and show it there, even when empty:

![screen shot 2017-10-26 at 09 39 13](https://user-images.githubusercontent.com/1204802/32040577-9dc44772-ba31-11e7-9319-e5b0bcbb1cbc.png)

This would allow us to change the scrollable areas top coordinate so it's never covered. I'm also confident that this toolbar area — when no block is selected — we'll soon find a use for. Whether through extensions or permalinks or something else.

jasmussen commented:
I also got this JS error while messing around, not sure if it's this branch or also master:

![screen shot 2017-10-26 at 09 36 56](https://user-images.githubusercontent.com/1204802/32040680-ea927588-ba31-11e7-83fb-51bad516af9b.png)

Try these steps to reproduce:

- Type in a paragraph
- In a new paragraph paste a bunch of text

Or this one:

- Type in a bunch of paragraphs
- Change one of them to a quote
- Delete the quote

youknowriad committed:
Block Switcher: Do not show if block not found

youknowriad commented:
I pushed a fix for the JS error 5d16fa3

youknowriad committed:
Block Toolbar: Always show on mobile

youknowriad commented:
The toolbar is always visible on mobile now.

jasmussen commented:
Nice. Couple of issues though:

![screen shot 2017-10-26 at 10 32 30](https://user-images.githubusercontent.com/1204802/32042836-3b5f3134-ba39-11e7-8616-8d9968a5865e.png)

It covers the sidebar

We'll also want to push down this region so the scrollbar starts after the docked toolbar:

![screen shot 2017-10-26 at 10 32 57](https://user-images.githubusercontent.com/1204802/32042847-48dbe398-ba39-11e7-90e1-e2b13d169b90.png)

Thanks for working on this.

youknowriad committed:
Polish Editor and Toolbar position in the different breakpoints

youknowriad commented:
Not easy to get right but I think it's ok now? bb70961

jasmussen commented:
This seems great to me now. And I do agree, would be good to get this in soon so we can both test it, and polish it before the next version. CC: @mtias

mtias commented:
@youknowriad is this rebased with master? I'm not seeing the keyboard improvements.

youknowriad commented:
@mtias not the last master, rebasing

youknowriad commented:
Merged to give us time to fix any missed bug before the release

jasmussen commented:
🔥 🔥 🔥 🔥 🔥
